From 17b45a7fa015e91570eb0509b5f7b8140975a88f Mon Sep 17 00:00:00 2001
From: AaronElijah <aaronzakelijah@googlemail.com>
Date: Wed, 2 Apr 2025 13:25:44 +0100
Subject: [PATCH] fix: mdelay to wait for peripherals to be ready before being
 probed

- Use board_init_* functions instead of device driver probes to make driver as general as possible
---
 .../switchbloxrugged_som.c                    |  3 ++-
 drivers/net/ksz9477.c                         | 22 ++++++++++++++++++-
 include/configs/switchbloxrugged_som.h        | 10 ++++++++-
 3 files changed, 32 insertions(+), 3 deletions(-)

diff --git a/board/botblox/switchbloxrugged_som/switchbloxrugged_som.c b/board/botblox/switchbloxrugged_som/switchbloxrugged_som.c
index 8105bf9040b..fa664627d4a 100644
--- a/board/botblox/switchbloxrugged_som/switchbloxrugged_som.c
+++ b/board/botblox/switchbloxrugged_som/switchbloxrugged_som.c
@@ -276,6 +276,7 @@ static inline int setup_lcd(void) { return 0; }
 
 int board_early_init_f(void)
 {
+	mdelay(20);   /* 20 ms needed for all peripeherals to be ready*/ 
 	return 0;
 }
 
@@ -349,7 +350,7 @@ int checkboard(void)
 	// 	puts("Board: MX6ULZ 14x14 EVK\n");
 	// else
 	// 	puts("Board: MX6ULL 14x14 EVK\n");
-	puts("Board: SwitchBloxRugged Rugged SOM");
+	puts("Board: BotBlox Rugged SOM\n");
 
 	return 0;
 }
diff --git a/drivers/net/ksz9477.c b/drivers/net/ksz9477.c
index 329724c95f7..1cfde050db2 100644
--- a/drivers/net/ksz9477.c
+++ b/drivers/net/ksz9477.c
@@ -168,9 +168,13 @@ static __maybe_unused void ksz_port_mmd_read(struct udevice *dev, int port,
 											 u8 addr, u16 reg, u16 *val)
 {
 	ksz_pwrite16(dev, port, REG_PORT_PHY_MMD_SETUP, MMD_SETUP(PORT_MMD_OP_INDEX, addr));
+	mdelay(1);
 	ksz_pwrite16(dev, port, REG_PORT_PHY_MMD_INDEX_DATA, reg);
+	mdelay(1);
 	ksz_pwrite16(dev, port, REG_PORT_PHY_MMD_SETUP, MMD_SETUP(PORT_MMD_OP_DATA_NO_INCR, addr));
+	mdelay(1);
 	ksz_pread16(dev, port, REG_PORT_PHY_MMD_INDEX_DATA, val);
+	mdelay(1);
 	dev_dbg(dev, "%s  P%d 0x%02x:0x%04x<<0x%04x\n", __func__, port + 1, addr, reg, *val);
 }
 
@@ -178,9 +182,13 @@ static void ksz_port_mmd_write(struct udevice *dev, int port, u8 addr, u16 reg,
 {
 	dev_dbg(dev, "%s P%d 0x%02x:0x%04x>>0x%04x\n", __func__, port + 1, addr, addr, val);
 	ksz_pwrite16(dev, port, REG_PORT_PHY_MMD_SETUP, MMD_SETUP(PORT_MMD_OP_INDEX, addr));
+	mdelay(1);
 	ksz_pwrite16(dev, port, REG_PORT_PHY_MMD_INDEX_DATA, addr);
+	mdelay(1);
 	ksz_pwrite16(dev, port, REG_PORT_PHY_MMD_SETUP, MMD_SETUP(PORT_MMD_OP_DATA_NO_INCR, addr));
+	mdelay(1);
 	ksz_pwrite16(dev, port, REG_PORT_PHY_MMD_INDEX_DATA, val);
+	mdelay(1);
 }
 
 /* Apply PHY settings to address errata listed in KSZ9477, KSZ9897, KSZ9896, KSZ9567
@@ -195,11 +203,14 @@ static void ksz_port_mmd_write(struct udevice *dev, int port, u8 addr, u16 reg,
  */
 static void ksz_phy_errata_setup(struct udevice *dev, int port)
 {
-	dev_dbg(dev, "%s P%d\n", __func__, port + 1);
+	dev_info(dev, "%s P%d\n", __func__, port + 1);
 
+	/* Disable auto-negotiation */
+	dev_dbg(dev, "Disabling auto-negotiation on port %d for errata fixes\n", port + 1);
 	ksz_pwrite16(dev, port, 0x0100, BMCR_SPEED100 | BMCR_FULLDPLX);
 
 	/* Register settings are needed to improve PHY receive performance */
+	dev_dbg(dev, "Register settings are needed to improve PHY receive performance to port %d\n", port + 1);
 	ksz_port_mmd_write(dev, port, 0x01, 0x6f, 0xdd0b);
 	ksz_port_mmd_write(dev, port, 0x01, 0x8f, 0x6032);
 	ksz_port_mmd_write(dev, port, 0x01, 0x9d, 0x248c);
@@ -209,12 +220,15 @@ static void ksz_phy_errata_setup(struct udevice *dev, int port)
 	ksz_port_mmd_write(dev, port, 0x1c, 0x08, 0x2001);
 
 	/* Transmit waveform amplitude can be improved (1000BASE-T, 100BASE-TX, 10BASE-Te) */
+	dev_dbg(dev, "Improving transmit waveform amplitude on port %d\n", port + 1);
 	ksz_port_mmd_write(dev, port, 0x1c, 0x04, 0x00d0);
 
 	/* Energy Efficient Ethernet (EEE) feature select must be manually disabled */
+	dev_dbg(dev, "Disabling Energy Efficient Ethernet (EEE) feature select on port %d\n", port + 1);
 	ksz_port_mmd_write(dev, port, 0x07, 0x3c, 0x0000);
 
 	/* Register settings are required to meet data sheet supply current specifications */
+	dev_dbg(dev, "Register settings are required to meet data sheet supply current specifications on port %d\n", port + 1);
 	ksz_port_mmd_write(dev, port, 0x1c, 0x13, 0x6eff);
 	ksz_port_mmd_write(dev, port, 0x1c, 0x14, 0xe6ff);
 	ksz_port_mmd_write(dev, port, 0x1c, 0x15, 0x6eff);
@@ -577,6 +591,10 @@ static int ksz_i2c_probe(struct udevice *dev)
 	data8 |= SW_RESET;
 	ksz_write8(priv->dev, REG_SW_OPERATION, data8);
 
+	int delay_dur = 200;
+	dev_info(dev, "%s: reset chip, delay %dms\n", __func__, delay_dur);
+	mdelay(delay_dur);   /* wait for chip reset */
+
 	/* read chip id */
 	ret = ksz_read32(dev, REG_CHIP_ID0__1, &id);
 	if (ret)
@@ -625,6 +643,8 @@ static int ksz_i2c_probe(struct udevice *dev)
 	/* Apply PHY errata by default */
 	for (i = 0; i < pdata->num_ports; i++) 
 	{
+		if (i == pdata->cpu_port)
+			continue;
 		ksz_phy_errata_setup(priv->dev, i);
 	}
 
diff --git a/include/configs/switchbloxrugged_som.h b/include/configs/switchbloxrugged_som.h
index b36f3aa9ade..a762d3c9fab 100755
--- a/include/configs/switchbloxrugged_som.h
+++ b/include/configs/switchbloxrugged_som.h
@@ -43,6 +43,13 @@
 #define MFG_NAND_PARTITION ""
 #endif
 
+// We don't want to use the inbuilt device driver for this as it is not necessary to configure PHY devices for the integrated ports on the switch 
+// Instead, we can write any configuration we want here and disable the KSZ9477/KSZ8567 driver in config
+// #define CFG_KSZ8567_I2C_SETTINGS \
+//     "i2c_prepare=i2c dev 0\0" \
+//     "ksz8567_i2c_addr=0x5f\0" \
+//     "ksz_check=i2c probe ${ksz8567_i2c_addr}\0" \
+
 #define CFG_MFG_ENV_SETTINGS       \
     CFG_MFG_ENV_SETTINGS_DEFAULT   \
     "initrd_addr=0x86800000\0"     \
@@ -86,9 +93,10 @@
     "fdt_file=imx6ull-botblox-swr-som.dtb\0"                                                                                 \
     "fdt_addr=0x83000000\0"                                                                                                  \
     "tee_addr=0x84000000\0"                                                                                                  \
-    "tee_file=uTee-6ullevk\0"                                                                                                 \
+    "tee_file=uTee-6ullevk\0"                                                                                                \
     "boot_fdt=try\0"                                                                                                         \
     "ip_dyn=yes\0"                                                                                                           \
+    "ipaddr=192.168.2.1\0"                                                                                                   \
     "splashimage=0x8c000000\0"                                                                                               \
     "mmcdev="__stringify(CONFIG_SYS_MMC_ENV_DEV) "\0"                                                                        \
                                                  "mmcpart=1\0"                                                               \
-- 
2.49.0

